### ISSUE TITLE
Hierarchical Logistic Regression Model Won't Converge?

### ISSUE THREAD

**Issue by jwdink** [greta-dev/greta#107]
Hi,

I'm back with another one of those annoying "it seems to work in Stan but not in greta-- am I doing something wrong?" questions.

The reason this one seemed worth raising an issue over is that hierarchical logistic regression is--at least, I think of it as-- kind of a bread and butter case for Bayesian MCMC methods (e.g., see https://arxiv.org/abs/1701.04858). I was surprised that greta was having so much trouble converging in this case, since there's nothing fancy going on here.

Here's the example:

```
library('dplyr')
# tried both of these:
#devtools::install_github("greta-dev/greta") 
#devtools::install_github("goldingn/greta@dev")
library('greta')
set.seed(2017-10-10)



# Make Data ---------------------------------------------------------------
num_subs <- 250
df_subs <- data_frame(id=1:num_subs,
                      num_trials = sample(10:150, size = num_subs, replace = TRUE))
df <- left_join(df_subs,
                bind_rows(lapply(unique(df_subs$num_trials), function(.x) data_frame(num_trials=.x, trial=1:.x)))) %>%
  select(-num_trials)
num_betas <- 10 # not including intercept
true_betas <- rnorm(num_betas+1, mean = 1)
true_betas[sample(num_betas, size=floor(num_betas)/2)+1] <- 0
for (i in 0:num_betas) {
  df_subs[[paste0("beta",i)]] <- true_betas[i+1] + rnorm(nrow(df_subs), sd = abs(true_betas[i+1])+.001 )
  if (i==0) df$X0 <- 1
  else df[[paste0("X",i)]] <- rnorm(nrow(df))
}
df$logodds <- rowSums(as.matrix(df_subs[df$id,-(1:2)]) * as.matrix(select(df, starts_with("X")))) + rnorm(nrow(df),sd = .5)
df$Y <- rbinom(n = nrow(df), size = 1, prob = plogis(df$logodds))
df$X0 <- NULL



# Greta Model -----------------------------------------------------------
compiled_model <- {
  X <- model.frame(reformulate(paste0("X",seq_len(num_betas))), data=df) %>%
    model.matrix(terms(.), data=.)
  Y <- df$Y
  group_inds <- df$id
  num_groups <- n_distinct(group_inds)
  num_obs <- nrow(X)
  num_betas <- ncol(X)
  
  betas <- greta::normal(0, 5, dim=c(1,ncol(X)))
  Omega <- greta::lkj_correlation(1,dim = num_betas)
  sigma <- greta::lognormal(0,1,dim=num_betas)
  sigma_mat <- greta_array(dim = c(num_betas,num_betas))
  diag(sigma_mat) <- sigma
  Sigma <- sigma_mat %*% Omega %*% sigma_mat
  #beta_offsets_raw <- greta::normal(0, 1, dim = c(num_groups, num_betas))
  #group_offsets_sd <- greta::lognormal(0,1, dim = c(1,num_betas))
  #beta_offsets <- group_offsets_sd[rep(1,num_groups),]*beta_offsets_raw
  beta_offsets <- greta::multivariate_normal(zeros(num_betas),Sigma, dim = num_groups)
  betas_per_group <- betas[rep(1,num_groups),] + beta_offsets
  yhat <- greta:::rowSums(betas_per_group[group_inds,] * X)
  
  distribution(Y) <- greta::bernoulli(prob = greta::ilogit(yhat))
  compiled_model <- greta::model(betas, Omega, sigma, precision = "double")
  compiled_model
}
#plot(compiled_model)
draws_list <- replicate(2, mcmc(compiled_model, warmup = 1000, n_samples = 1000, pb_update = 50))
              #,initial_values = rep(0.01, length=length(compiled_model$dag$example_parameters())) 
draws <- coda::mcmc.list(draws_list)
diagn <- sirt::mcmc.list.descriptives(draws)[,'Rhat',drop=FALSE]
head(round(diagn,2), num_betas+1)
```

Even though it seems to roughly recover the true parameters, convergence diagnostics suggest something is amiss, so in a real setting with unknown parameter-values, I'd be wary. (I also tried independent normals instead of multivariate-- see the commented out code. This helps some but not enough.)

Assuming I'm doing this correctly, this is the canonical setup for a hierarchical regression model. So it should be pretty much equivalent to the following:

```
library('rstanarm')
form <- reformulate(paste0("X",seq_len(num_betas)), response = "Y")
form <- update(form, .~.+(1|id))
form[[3]][[3]][[2]][[2]] <- reformulate(paste0("X",seq_len(num_betas)))[[2]]
fit_stan <- stan_glmer(formula = form,
                       data = df,
                       cores = 2, refresh=10,
                       family = stats::binomial)
fit_stan$stan_fit
```

Which doesn't seem to have convergence issues.

**Comment by goldingn** [comment]
Hi @jwdink,

Looks like it isn't converging well because the multivariate normal prior on `beta_offsets` induces quite strong posterior correlations in those variables, and those correlations are variable (depending on the values of `Omega` at that iteration).

There's a nice reparameterization trick, the [whitening transformation](https://en.wikipedia.org/wiki/Whitening_transformation) to resolve this problem of variables with known correlation structure: we can change the definition of the multivariate normal vector `x` (with length `n`) from:
```r
x = multivariate_normal(mu, Sigma, dim = n)
```
to
```r
z = normal(0, 1, dim = n)
x = mu +  z %*% chol(Sigma)
```
`x` has the same distribution, but in the latter, the sampler is looking at the variables `z`, which aren't subject to the correlations or variances defined in `Sigma`. This trick is used in rstanarm too, though [the code](https://github.com/stan-dev/rstanarm/blob/b78c2b5190db8b62da93f0c686d0c78da4e5bb9b/inst/chunks/common_functions.stan#L77-L120) for it is pretty opaque.

This is very similar to the non-centred reparameterisation for univariate hierarchical models (the Matt trick in stan parlance) to avoid nasty funnel effects in sampling. I am considering making the non-centred parameterisation happen automatically when using `normal()` in greta. I'll also consider making whitening automatic for `multivariate_normal()`.

For now, you just need to replace your line: 
```r
beta_offsets <- greta::multivariate_normal(zeros(num_betas),Sigma, dim = num_groups)
```
with
```r
beta_offsets_raw <- greta::normal(0, 1, dim = c(num_groups, num_betas))
beta_offsets <- beta_offsets_raw %*% chol(Sigma)
```

That brings all those Rhats to ~1 for me.

---
P.S. you can actually skip the cholesky decomposition altogether by being clever with the underlying representation of `Omega`. Though it makes for hideous code here and doesn't provide much of a speed up for such a small matrix. If there were 100s of covariates it would be worth doing. I plan to make those tricks automatic too.

**Comment by jwdink** [comment]
Wow, very nice-- thank you. 

I actually am hoping to use this in a setting with a larger dataset and more predictors (potentially with a laplace prior on the betas instead of a normal prior, assuming that behaves well). So I would love to know more about the 'clever trick' with Omega. 

Doubly so because, while the reparameterization above definitely helps, I'm still seeing r-hats that up in the 1.2-2.0+ range, which I can't seem to get rid of by increasing iterations. I imagine this will get worse with more predictors. Would the Omega-trick help there? Or anything else I can do?

**Comment by goldingn** [comment]
Sure thing. Basically correlation matrix (e.g. created with `lkj_correlation()`) and covariance matrix variables (e.g. created with `wishart()`) are actually modelled via their Cholesky factors under the hood. I.e. the non-zero elements of those triangular matrices are flattened into a vector and inference is performed on that transformation during HMC.

For subsequent operations, like computing the multivariate normal log density, that Cholesky factor is what's needed so greta will skip the matrix multiplication and cholesky decomposition steps. I have plans to use the same mechanism to skip more redundant calculations (#86) but that's not yet implemented.

In this case, you can dig out the cholesky factor of `Omega` (converting it from a node to a greta array), and add the marginal variances directly to the cholesky factor to get `U`; the cholesky factor of `Sigma`:

```r
Omega_U <- greta:::as.greta_array(Omega$node$representations$cholesky_factor)
U <- sweep(Omega_U, 2, sigma, '*')
```
and use `U` where you use `chol(Sigma)` above:
```r
beta_offsets <- beta_offsets_raw %*% U
```

This should be more efficient for lots of covariates, but it probably won't affect convergence. I'm surprised to hear you're still getting high Rhats, they're all close to 1 when I run the example above. Are you having those problems on a different or larger example?

**Comment by jwdink** [comment]
Hmmm; when I run the original code w/seed, but with the modification for the whitening transformation, I get:

```
> diagn %>% tibble::rownames_to_column() %>% count(round(Rhat,1)) %>% knitr::kable()

| round(Rhat, 1)|  n|
|--------------:|--:|
|            1.0| 55|
|            1.1| 42|
|            1.2| 16|
|            1.3|  9|
|            1.4|  2|
|            1.5|  2|
|            1.6|  3|
|            1.7|  4|
|            1.8|  2|
|            2.2|  2|
|            5.5|  2|
|            6.0|  2|
|            NaN|  2|
```

The rule of thumb is we want all under 1.1, right?

Thank you for the tip for improved performance w/more covariates!

**Comment by goldingn** [comment]
Hi @jwdink,

Sorry I let this issue go stale! I've been planning for a while to dig into convergence issues with greta, but haven't had the time. I've recently made some progress speeding up the HMC sampler (by importing raw TensorFlow samplers from the new TensorFlow Probability library) and improving how greta tubes the sampler hyperparameters. There's a development version that you can install with `devtools::install_github("goldingn/greta@tweak_tuning")`. You may need to install more stuff, but it should prompt you what to do.

I realise you've probably moved on with this model, but if you have code for the bigger model you mentioned (or any other model) that wasn't converging nicely before, I'd be very keen for test cases to shore up the new tuning regime!

**Comment by goldingn** [comment]
I'm pretty happy with convergence and tuning on the dev branch now, and models like this seem to run fine, so I'm considering this closed. Please feel free to open a new issue if there are convergence problems with other models with the dev version
